File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void allCombs(int idx, string &digs, string &temp, vector<string> &v, vector<string> &ans){
        if(idx==digs.size()){
            ans.push_back(temp);
            return ;
        }
        for(int i=0; i<v[(digs[idx]-'0')-2].size(); i++){
            temp.push_back(v[(digs[idx]-'0')-2][i]);
            allCombs(idx+1, digs, temp, v, ans);
            temp.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0) return {};
        vector<string> v= {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> ans;
        string temp="";
        allCombs(0, digits, temp, v, ans);
        return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    private void helper(int idx, String digs, StringBuffer temp, String[] pad, List<String> ans){
        if(idx== digs.length()){
            ans.add(temp.toString());
            return;
        }
        for(int i=0; i<pad[digs.charAt(idx)-'0' -2].length(); i++){
            temp.append(pad[digs.charAt(idx)-'0' -2].charAt(i));
            helper(idx+1, digs, temp, pad, ans);
            temp.deleteCharAt(temp.length()-1);
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<String>();
        if(digits.length()==0) return ans;
        StringBuffer temp = new StringBuffer();
        String []pad = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
        helper(0, digits, temp, pad, ans);
        
        return ans;
    }
}